src/main/java/com/google/javascript/clutz/AliasMapBuilder.java

```diff
@@ -49,43 +50,93 @@
-            variableName = originalName;
+            localVariableName = originalName;
           }
 
-          String exportedSymbolName = buildNamedExportSymbolName(importedModuleId, originalName);
-          aliasMap.putAll(
-              buildNamedAndWholeModuleMappings(localModuleId, variableName, exportedSymbolName));
+          String importedSymbolName = buildNamedExportSymbolName(importedModuleId, originalName);
+          localVariableToImportedSymbolNameMap.put(localVariableName, importedSymbolName);
+        }
+      }
+    }
+
+    // Loop over the statements, looking for export statements, and add mappings to the alias map
+    // if the export is of a variable that was imported
+    for (Node statement : moduleBody.children()) {
+      if (isWholeModuleExportAssignment(statement)) {
+        // `exports = foo`
+        String localVariableName = statement.getFirstChild().getChildAtIndex(1).getString();
+
+        if (localVariableToImportedSymbolNameMap.containsKey(localVariableName)) {
+          aliasMap.put(
+              buildWholeModuleExportSymbolName(localModuleId),
+              localVariableToImportedSymbolNameMap.get(localVariableName));
+        }
+      } else if (isNamedExportAssignment(statement)) {
+        // `exports.foo = foo`
+        String localVariableName = statement.getFirstChild().getChildAtIndex(1).getString();
+        String exportName =
+            statement.getFirstChild().getFirstChild().getChildAtIndex(1).getString();
+
+        if (localVariableToImportedSymbolNameMap.containsKey(localVariableName)) {
+          aliasMap.put(
+              buildNamedExportSymbolName(localModuleId, exportName),
+              localVariableToImportedSymbolNameMap.get(localVariableName));
         }
       }
-      // TODO(lucassloan): actually parse the `exports = foo` or `exports.foo = foo` statements
-      // to avoid overwriting exports that aren't actually aliases.
     }
 
     return aliasMap;
   }
 
-  /**
-   * To avoid parsing for the export statements, generate mappings for both styles of exports from
-   * goog.modules.
-   *
-   * <p>`exports = Foo` produces the name `module$exports$module$Name` `exports.Foo = Foo` produces
-   * the name `module$exports$module$Name.Foo`
-   */
-  private static Map<String, String> buildNamedAndWholeModuleMappings(
-      String localModuleId, String variableName, String exportedSymbolName) {
-    //TODO(lucassloan): handle knownGoogProvides
-    Map<String, String> mappings = new HashMap<>();
-    // exports = Foo
-    mappings.put(buildWholeModuleExportSymbolName(localModuleId), exportedSymbolName);
-    // exports.Foo = Foo
-    mappings.put(buildNamedExportSymbolName(localModuleId, variableName), exportedSymbolName);
-
-    return mappings;
+  /** Matches `exports = foo;` */
+  protected boolean isWholeModuleExportAssignment(Node statement) {
+    if (!statement.isExprResult()) {
+      return false;
+    }
+
+    if (!statement.getFirstChild().isAssign()) {

```

Comment:
Previous such code in gents, ended up with NPEs, because some closure apis can return null. Can you check for null, or convince me that it can't happen for things like getFirstChild() and getChildAtIndex(1).

Reply:
I don't think we're vulnerable to NPEs here, since we're checking for the types of the parent node before we access the child nodes.  ImportBasedMapBuilder has a bunch of similar checks which work just fine without explicit null checks.